### ISSUE TITLE
Serialisation exception with MySQL session store

### ISSUE THREAD

**Issue by ghalse** [simplesamlphp/simplesamlphp#399]
TL;DR - TEXT is too small for kvstore in MySQL, LONGTEXT works.

We use MySQL as a data store, and store sessions in SQL. There are certain SP entities we simply cannot log into - we never get beyond the login screen, and each time we authenticate (successfully), we're returned to the same login screen. Other SP entities work just fine.

The logs show a backtrace similar to this each time:

Jun  1 09:02:43 idp-01 simplesamlphp[4989]: 3 [f826865b38] SimpleSAML_Error_Exception: Error 8 - unserialize(): Error at offset 528 of 49337 bytes
Jun  1 09:02:43 idp-01 simplesamlphp[4989]: 3 [f826865b38] Backtrace:
Jun  1 09:02:43 idp-01 simplesamlphp[4989]: 3 [f826865b38] 8 /srv/simplesamlphp-1.14.2/www/_include.php:75 (SimpleSAML_error_handler)
Jun  1 09:02:43 idp-01 simplesamlphp[4989]: 3 [f826865b38] 7 [builtin] (unserialize)
Jun  1 09:02:43 idp-01 simplesamlphp[4989]: 3 [f826865b38] 6 /srv/simplesamlphp-1.14.2/lib/SimpleSAML/Store/SQL.php:253 (SimpleSAML_Store_SQL::get)
Jun  1 09:02:43 idp-01 simplesamlphp[4989]: 3 [f826865b38] 5 /srv/simplesamlphp-1.14.2/lib/SimpleSAML/SessionHandlerStore.php:52 (SimpleSAML_SessionHandlerStore::loadSession)
Jun  1 09:02:43 idp-01 simplesamlphp[4989]: 3 [f826865b38] 4 /srv/simplesamlphp-1.14.2/lib/SimpleSAML/Session.php:255 (SimpleSAML_Session::getSession)
Jun  1 09:02:43 idp-01 simplesamlphp[4989]: 3 [f826865b38] 3 /srv/simplesamlphp-1.14.2/lib/SimpleSAML/Session.php:192 (SimpleSAML_Session::getSessionFromRequest)
Jun  1 09:02:43 idp-01 simplesamlphp[4989]: 3 [f826865b38] 2 /srv/simplesamlphp-1.14.2/lib/SimpleSAML/Auth/State.php:253 (SimpleSAML_Auth_State::loadState)
Jun  1 09:02:43 idp-01 simplesamlphp[4989]: 3 [f826865b38] 1 /srv/simplesamlphp-1.14.2/modules/consent/www/getconsent.php:33 (require)
Jun  1 09:02:43 idp-01 simplesamlphp[4989]: 3 [f826865b38] 0 /srv/simplesamlphp-1.14.2/www/module.php:127 (N/A)

On investigation, I discovered that the one thing that all the entities that did not work had in common is that they were created using the metarefresh module. However not all entities created via metarefresh have the same problem.

The common factor is that all the entities that do have a problem have multiple X.509 certificates, usually quite large ones. Yet these entities work fine if I add them by hand, rather than using metarefresh. The difference between the manually added metadata and the metadata created by metarefresh is the entityDescriptor key, which contains a base64 representation of the original XML.

By default in MySQL the kvstore is created with _value as a TEXT type. This can store 2^16 bytes. Given the 49337 number above, this made me wonder if the problem was related to the size of the field in the SQL datastore being exceeded - and thus the serialised session state being truncated, causing the above error when it is retrieved.

As a test, I did the following:

ALTER TABLE SimpleSAMLphp_kvstore MODIFY _value LONGTEXT;

on my database, and tried the login again. This now completes without any problems for all entities I've tested with, which seems to confirm my suspicion than the TEXT field size is simply too small.

**Comment by ghalse** [comment]
Simplistically this can be fixed by changing the CREATE statement in SimpleSAML_Store_SQL::initKVTable(). However I don't know enough about how this would affect other SQL variants.

**Comment by jaimeperez** [comment]
Hi @ghalse!

Thanks for reporting this issue. Unfortunately, it appears like the `LONGTEXT` data type is specific to MySQL, so we can't simply switch to it when creating the table, as other backends I've just verified (PostgreSQL, Oracle, MS SQL) don't have it.

Instead, I've committed a fix that would change the data type if MySQL is used. Hopefully, that will prevent this issue from happening again in new installations.

Thanks again!
